File: 03_Sliding_Window/04_Permutation_in_String/0567-permutation-in-string.py
"""
Problem: LeetCode 567 - Permutation in String

Key Idea:
To check whether a string 's2' contains a permutation of another string 's1', we can use a sliding window approach. First, we create a frequency dictionary for characters in 's1'. Then, we initialize two pointers, 'left' and 'right', to represent the current window in 's2'. As we move the 'right' pointer to the right, we add the character to a temporary frequency dictionary and check if it becomes equal to the frequency dictionary of 's1'. If it does, it means we found a permutation of 's1' in 's2', and we return True. If the window size exceeds the length of 's1', we remove the character at the 'left' pointer from the temporary dictionary and move the 'left' pointer to the right to shrink the window. We continue this process until we find a permutation or reach the end of 's2'.

Time Complexity:
The time complexity of this solution is O(n), where n is the length of 's2'. The sliding window approach iterates through 's2' once, and at each step, we perform constant-time operations to update the window and the frequency dictionaries.

Space Complexity:
The space complexity is O(1) since the frequency dictionaries have a constant number of elements (the number of unique characters in 's1' or the alphabet size) regardless of the size of the input strings 's1' and 's2'.
"""


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        char_freq_s1 = {}
        for char in s1:
            char_freq_s1[char] = char_freq_s1.get(char, 0) + 1

        left, right = 0, 0
        char_freq_temp = {}

        while right < len(s2):
            char_freq_temp[s2[right]] = char_freq_temp.get(s2[right], 0) + 1

            if right - left + 1 == len(s1):
                if char_freq_temp == char_freq_s1:
                    return True
                char_freq_temp[s2[left]] -= 1
                if char_freq_temp[s2[left]] == 0:
                    del char_freq_temp[s2[left]]
                left += 1

            right += 1

        return False

File: 03_Sliding_Window/04_Permutation_in_String/0567-permutation-in-string.rs
/*
Problem: LeetCode 567 - Permutation in String

Key Idea:
The key idea is to use a sliding window approach with two frequency maps. Create a frequency map of characters in the pattern string, then slide a window of the same length over the main string and maintain a frequency map of characters within the window. Compare the two frequency maps at each step to check for a permutation.

Approach:
1. Initialize two arrays, `target_freq` and `window_freq`, each of size 26 to represent the frequency of characters in the target string and the current window, respectively.
2. Convert the input strings to byte arrays to work with ASCII values.
3. Initialize two pointers, `left` and `right`, representing the start and end of the current window.
4. Iterate through the `s2` string using the `right` pointer:
   - Update the frequency of the character at position `right` in the `window_freq` array.
   - If the window size is greater than or equal to the length of `s1`, decrement the frequency of the character at position `left` in the `window_freq` array and move the `left` pointer to the right.
   - If the frequencies of characters in `window_freq` match the frequencies in `target_freq`, it means a permutation of `s1` is found.
5. Return `true` if a permutation is found, otherwise return `false`.

Time Complexity:
O(n), where n is the length of the `s2` string. We perform a single pass through the `s2` string using the sliding window approach.

Space Complexity:
O(1), as the arrays `target_freq` and `window_freq` are of constant size 26.
*/

impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let s1 = s1.as_bytes();
        let s2 = s2.as_bytes();
        let mut target_freq = [0; 26];
        let mut window_freq = [0; 26];

        for i in 0..s1.len() {
            target_freq[(s1[i] - b'a') as usize] += 1;
            window_freq[(s2[i] - b'a') as usize] += 1;
        }

        let mut left = 0;

        for right in s1.len()..s2.len() {
            if window_freq == target_freq {
                return true;
            }

            window_freq[(s2[left] - b'a') as usize] -= 1;
            left += 1;
            window_freq[(s2[right] - b'a') as usize] += 1;
        }

        window_freq == target_freq
    }
}
